### ISSUE TITLE
delayed reaction on F* keys just after launch

### ISSUE THREAD

**Issue by unxed** [elfmz/far2l#1030]
А это только у меня в последних версиях far2l сразу после запуска при нажатии F1..F12 реакция происходит с задержкой в секунду-полторы? Только в первый раз после запуска, потом моментальная реакция. Только в wx-версии ловил, в консольной ок всё.

**Comment by elfmz** [comment]
У меня норм, а версия не с питоном ли собрана? Изза похожих (но не уверен именно насчет F*) эффектов я питон и не включаю по дефолту.

**Comment by unxed** [comment]
> У меня норм, а версия не с питоном ли собрана? Изза похожих (но не уверен именно насчет F*) эффектов я питон и не включаю по дефолту.

с питоном! спасибо, сейчас без него перепроверю

**Comment by unxed** [comment]
да, без питона проблема исчезла. выходит, надо тэгнуть автора питонового плагина?

**Comment by unxed** [comment]
In the latest versions of far2l with python plugin enabled, when you press F1..F12 immediately after starting far2l, the reaction occurs with a delay of one and a half seconds. Only the first time after launch. Only in the wx-version, in console everything is ok.

@m32

**Comment by elfmz** [comment]
I suspect its an initialization of python and can't be simple avoided. However may think about background initialization in separate thread, also this can be nontrivial to implement.

**Comment by m32** [comment]
The python plugin uses cffi, which compiles files at runtime. Static plugin generation should speed up the loading process. I will see if this process can be accelerated.

**Comment by elfmz** [comment]
Is it possible to store compilation results into files and load them later as long sources didnt change?

**Comment by m32** [comment]
Statically generating the library did nothing. Only wrappers for constants are generated, no support for procedures and pointers - cffi is inherently dynamic vs pybind, swig, ...

The time to run the setup procedure in pluginmanager.py on my laptop is about 0.0018303 seconds, and the response to pressing any key is immediate. What we're writing about here may be about loading the python interpreter itself, and the python plugin has no effect on that. It's just that python loads a lot of libraries and its start is not very fast :(, e.g. lua is much faster.

**Comment by unxed** [comment]
Running python3 from the command line happens instantly for me

**Comment by elfmz** [comment]
It seems delay caused by this line pyPluginModule = PyImport_Import(pName). See, i caught a stack at the moment of delay:
Thread 8 (Thread 0x7fffda925700 (LWP 5165)):
#0  0x00007fffd8a20ed9 in _PyType_Lookup () at /usr/lib/x86_64-linux-gnu/libpython3.6m.so.1.0
#1  0x00007fffd8a3417b in _PyObject_GenericGetAttrWithDict () at /usr/lib/x86_64-linux-gnu/libpython3.6m.so.1.0
...87 lines skipped...
#88 0x00007fffd8a8e768 in PyObject_Call () at /usr/lib/x86_64-linux-gnu/libpython3.6m.so.1.0
#89 0x00007fffd8a8f058 in PyObject_CallFunction () at /usr/lib/x86_64-linux-gnu/libpython3.6m.so.1.0
#90 0x00007fffd898307b in PyImport_Import () at /usr/lib/x86_64-linux-gnu/libpython3.6m.so.1.0
#91 0x00007fffd8eed969 in SetStartupInfoW () at /home/user/projects/far2l.build/install/Plugins/python/plug/python.far-plug-wide
#92 0x0000555555696894 in PluginW::SetStartupInfo(bool&) [clone .part.25] ()
#93 0x00005555556968d3 in PluginW::SetStartupInfo(bool&) ()

**Comment by m32** [comment]
Incemetal execution time in milliseconds:
startup 0,00
dlopen 0,10
Py_SetProgramName 0,14
Py_Initialize 24,41
PyEval_InitThreads 24,48
PyInit__pyfar 25,32
flog 25,43
PyRun_SimpleString 25,51
PyUnicode_DecodeFSDefault 25,55
PyImport_Import 600,35
PyObject_GetAttrString 600,43
vcall 633,67

total time 633,74

PyImport_Import does the pluginmanager import and compiles far headers but that doesn't take 1500ms (for me)
In other words 1500ms is likely and very computer configuration dependent (I have SSDs and free RAM)

**Comment by unxed** [comment]
8 Gb RAM, SSD, AMD A10-4600M, Linux Mint Mate x64, no other apps running. Okay, it could have been a second, not one and a half - I did not take strict measurements.

**Comment by elfmz** [comment]
Please check branch async-python-init but check not only for discussed issue, but verify also if python still functioning properly :)
What i did there - is added special flag telling far2l to load python plugin immediately on start but invoke its exports lazily - likewise all other 'cached' plugins. And used this time gap between plugin load and any event that causes its use to load python stuff in parallel thread.

**Comment by m32** [comment]
py:load ucharmap upanel, udialog - py:unload ... - everything showed up fine, compilation also without problems

**Comment by elfmz** [comment]
Fine, for me issue with delay on first even almost eliminated. Still there is minor delay when 1st time python used, but much-much less than it was.

**Comment by elfmz** [comment]
Also suspicious thing i noticed - when started from terminal far2l with enabled python plugin cannot be terminated by Ctrl+C pressed in terminal - seems python somehow intercepts this signal. This worries cuz TTY version of far2l depend on signals processing to correctly handle terminal detachments etc. Will check later if this functionality works well with Python enabled, but may be you know how to control this signals processing by Python?

**Comment by m32** [comment]
Standard python has ctrl+c support - that's the KeyboardInterrupt exception.
I haven't turned it off anywhere so support should be turned on and if there is python support first (last installed sighandler) it will probably call python own routines first and won't let it go any further.

The answer on how to disable ctrl + c support should be somewhere in the wxpython source, I will look for after my vacation.

How and where far2l uses SIGINT (ctrl + c)?

**Comment by elfmz** [comment]
far2l doesnt use SIGINT, but it uses SIGHUP to detect terminal detachment. Also there is special handling of SIGSEGV for mmap'ed files and for crash reporting. If Python doesn't mess with that signals, but only only on SIGINT then everything should be fine.

**Comment by unxed** [comment]
Shouldn't we keep this open until Ctrl+C issues are resolved, or open an new issue instead?

**Comment by elfmz** [comment]
Nope, there're no issues with Ctrl+C just were suspicion that if SIGINT is intercepted by python then other signals can be also intercepted. Seems that suspicion didnt confirm.
